lib/routes/token.js

```diff
@@ -50,58 +166,37 @@ module.exports = {
     }
   },
   response: {
-    schema: {
-      access_token: Joi.string().required(),
+    schema: Joi.object().keys({
+      access_token: validators.token.required(),
+      refresh_token: validators.token,
       scope: Joi.string().required().allow(''),
       token_type: Joi.string().valid('bearer').required(),
-      auth_at: Joi.number().required()
-    }
+      expires_at: Joi.number().required(),
+      // you only can get an auth_at value if the grant_type was a CODE
+      // we cant check that here, but we can check another invariant:
+      // if this grant was with a code, then we ARE returning a
+      // refresh_token. So, we can say that refresh_token and auth_at
+      // comes as a pair.
+      auth_at: Joi.number(),
+    }).and('refresh_token', 'auth_at')
   },
   handler: function tokenEndpoint(req, reply) {
     var params = req.payload;
-    db.getClient(buf(params.client_id)).then(function(client) {
-      if (!client) {
-        logger.debug('client.notFound', { id: params.client_id });
-        throw AppError.unknownClient(params.client_id);
-      }
-
-      var submitted = hex(encrypt.hash(buf(params.client_secret)));
-      var stored = hex(client.secret);
-      if (submitted !== stored) {
-        logger.info('client.mismatchSecret', { client: params.client_id });
-        logger.verbose('client.mismatchSecret.details', {
-          submitted: submitted,
-          db: stored
+    confirmClient(params.client_id, params.client_secret).then(function() {
+      if (params.grant_type === AUTHORIZATION_CODE) {
+        return confirmCode(params.client_id, params.code).then(function(vals) {
+          vals.generateRefreshToken = true;
+          vals.ttl = params.ttl;
+          return vals;
         });
-        throw AppError.incorrectSecret(params.client_id);
-      }
-
-      return client;
-    })
-    .then(function(client) {
-      return db.getCode(Buffer(req.payload.code, 'hex')).then(function(code) {
-        if (!code) {
-          logger.debug('code.notFound', { code: req.payload.code });
-          throw AppError.unknownCode(req.payload.code);
-        } else if (String(code.clientId) !== String(client.id)) {
-          logger.debug('code.mismatch', {
-            client: client.id.toString('hex'),
-            code: code.clientId.toString('hex')
-          });
-          throw AppError.mismatchCode(req.payload.code, client.id);
-        } else {
-          // + because loldatemath. without it, it does string concat
-          var expiresAt = +code.createdAt + config.get('expiration.code');
-          if (Date.now() > expiresAt) {
-            logger.debug('code.expired', { code: code });
-            throw AppError.expiredCode(req.payload.code, expiresAt);
-          }
-        }
-        return code;
-      });
+      } else if (params.grant_type === REFRESH_TOKEN) {
+        return confirmToken(params).then(function(vals) {
+          vals.ttl = params.ttl;
```

Comment:
I know it's unnecessary, but I'd consider putting an explicit `vals.generateRefreshToken = false` here for completeness.